### PR TITLE
Add the capability of replacing rust functions with coprocessor lookups

### PR DESCRIPTION
**Pull request by gballet**
This PR implement a dummy `poseidon_hash` function, and substitutes it with a `poseidon` instruction in powdr assembly. This instruction is currently a stub that has to be replaced with a call to the actual poseidon coprocessor when #395 has been merged.

TODO: 

- [x] since #375 has been merged, this might need to be rebased in order to capture the call in the LLVM layer
- [x] ~~Build the sponge in rust~~ -> @leonardoalt said to leave it for another PR
- [x] The reachability code does not seem to successfully remove the translated `poseidon_hash` function, even though it's not called.
- [x] Geth @georgwiese's input on the parameters for the poseidon hash coprocessor and what the signature for the dummy function should be
  - [x]  In the meantime, find a better way of doing dummy code, so that the constraint checker doesn't crash.

### CODE REVIEW

File: riscv/runtime/src/coprocessors.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// This is a dummy implementation of Poseidon hash,
+// which will be replaced with a call to the poseidon
+// coporocessor during compilation.
+// The function itself will be removed by the compiler
+// during the reachability analysis.
+#[no_mangle]
+#[inline(never)]
+pub fn poseidon_hash(_data: &[u8], _size: usize) -> [u8; 32] {
+    return [0; 32];
+}
```
Comment (gballet) [inline riscv/runtime/src/coprocessors.rs:10]: This function's signature is inspired from the what the keccack256 one would be, but poseidon seems way more generic than that. I would appreciate some input here.

File: riscv/src/compiler.rs [hunk 2]
```diff
@@ -383,6 +383,12 @@ instr is_positive X -> Y {
     Y = wrap_bit
 }
 
+instr poseidon Y, Z -> X {
+    // Dummy code, to be replaced with actual poseidon code.
+    X = Y,
+    X = Z
```
Comment (gballet) [inline riscv/src/compiler.rs:389]: This is obviously a placeholder, but a very dumb one that doesn't let me check that everything works past the witness generation stage. Some input on how to write a better dummy would be appreciated.

For reference, this is what I get right now:

```
Error: Row 643: Identity check failed or unable to derive values for some witness columns.
Set RUST_LOG=debug for more information.
thread 'main' panicked at 'Witness generation failed.', /home/gballet/src/powdr/executor/src/witgen/generator.rs:189:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: riscv/src/compiler.rs [hunk 3]
```diff
@@ -1014,10 +1020,13 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
...
         "call" => {
-            if let [label] = args {
-                vec![format!("call {};", argument_to_escaped_symbol(label))]
-            } else {
-                panic!()
+            // Depending on what symbol is called, the call is replaced by a
+            // powdr asm call, or a call to a coprocessor if a special function
+            // has been recognized.
+            match args {
+                [Argument::Expression(Expression::Symbol(label))] if label == "poseidon_hash" => vec!["x0 <=X= poseidon(x10, x11);".to_string()],
```
Comment (gballet) [inline riscv/src/compiler.rs:1027]: it seems that powder assembly uses `x10` and `x11` as the first two parameters of its calling convention. Let me know if this assumption is incorrect.

**Review by gballet — COMMENTED** [review]
I have some open questions that I left inline.

File: riscv/runtime/src/coprocessors.rs (hunk 1 above)
Comment (georgwiese) [reply riscv/runtime/src/coprocessors.rs:10]: So in #395 (not merged yet, but working already), I implemented the basic Poseidon permutation (basically the $f$ [here](https://affine.group/images/2020/sponge.png). This lets you hash 2 field elements to one. There is an additional input (the capacity) which should be set to a constant, but it can be used for domain separation, so I think it could be exposed as well!

This already lets you build Merkle trees, which is a big use-case. The full Sponge construction is not implemented yet.

About going from bytes to field elements: How many bytes we can pack into one field element depends on the field. But I think it should be safe to assume that we can pack at least 32 bits (i.e., one RISC-V word).

Comment (georgwiese) [reply riscv/runtime/src/coprocessors.rs:10]: (It's working on the `poseidon2` branch, but only because it includes the changes from #393, so it wouldn't work yet on your branch).

Comment (gballet) [reply riscv/runtime/src/coprocessors.rs]: Thank you. @leonardoalt and I talked about this yesterday and we agreed that I will implement the sponge in rust for the time being, so that I can use your 2 -> 1 coprocesso.

Comment (georgwiese) [reply riscv/runtime/src/coprocessors.rs]: Cool! I'm not sure how it would work exactly though. In the case of BN254, the Poseidon permutation maps 3 field elements (rate `r1` & `r2`, capacity `c`) to 3 field elements.

In my understanding, the full Sponge construction should be:
```
r1, r2, c = 0, 0, 0
for m1, m2 in padded_message_blocks:
  r1, r2, c = poseidon_permutation(r1 + m1, r2 + m2, c)
return r1
```

In that case, you'd need all 3 inputs and outputs, right? (Also, in theory, you'd have to implement field addition in Rust...) I guess with 2 inputs and 1 output you could implement the Merkle–Damgård construction.

Anyway, all this is to say, let me know if you more inputs / outputs :)

File: riscv/runtime/src/coprocessors.rs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// This is a dummy implementation of Poseidon hash,
+// which will be replaced with a call to the poseidon
+// coporocessor during compilation.
+// The function itself will be removed by the compiler
+// during the reachability analysis.
+#[no_mangle]
```
Comment (chriseth) [inline riscv/runtime/src/coprocessors.rs:6]: I'm wondering if declaring this as a real FFI function would be better:
```
extern "C" {
    pub fn poseidon_hash(a: u32, b: u32) -> u32;
}
```

File: riscv/src/compiler.rs [hunk 5]
```diff
@@ -1128,10 +1135,15 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
             vec![format!("jump_and_link_dyn {rs};")]
         }
         "call" => {
```
Comment (chriseth) [inline riscv/src/compiler.rs:1137]: We should also at least support `tail`

File: riscv/src/compiler.rs [hunk 6]
```diff
@@ -1128,10 +1135,15 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
...
-            if let [label] = args {
-                vec![format!("call {};", argument_to_escaped_symbol(label))]
-            } else {
-                panic!()
+            // Depending on what symbol is called, the call is replaced by a
+            // powdr asm call, or a call to a coprocessor if a special function
+            // has been recognized.
+            match args {
+                [Argument::Expression(Expression::Symbol(label))] if label == "poseidon_hash" => {
```
Comment (chriseth) [inline riscv/src/compiler.rs:1142]: I think it would be better to make this configurable - the compiler should not need to know about which such functions exist and how they are implemented.

**Comment by leonardoalt** [comment]
> provide the PIL code to implement the function inside rust:

I currently disagree with that. It's ultra-generic and IMO we don't need that right now. I think it would be better to just define a map and some sort of "powdr co-processor stdlib" that can be used in Rust.

File: riscv/src/compiler.rs (hunk 5 above)
Comment (gballet) [reply riscv/src/compiler.rs:1141]: did that

File: riscv/runtime/src/coprocessors.rs (hunk 4 above)
Comment (gballet) [reply riscv/runtime/src/coprocessors.rs]: just pushed a commit that does that

File: riscv/src/compiler.rs (hunk 6 above)
Comment (gballet) [reply riscv/src/compiler.rs]: that makes sense, and I completely agree. I have just pushed another commit that does that, but it requires the experimental `if_let_guard` and it's still located in the compiler crate at the moment.

Comment (gballet) [reply riscv/src/compiler.rs]: which breaks the tests in CI :(

File: riscv/src/compiler.rs [hunk 7]
```diff
@@ -1128,10 +1173,17 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
...
+            // powdr asm call, or a call to a coprocessor if a special function
+            // has been recognized.
+            match args {
+                [Argument::Expression(Expression::Symbol(label))]
+                    if let Some(replacement) = has_coprocessor_substitution(label) =>
+                {
+                    vec![replacement.to_string()]
+                }
+                [label] => vec![format!("call {};", argument_to_escaped_symbol(label))],
+                _ => panic!(),
```
Comment (gballet) [inline riscv/src/compiler.rs:1186]: the `if let` guard is why I have to add a nightly feature and so why CI breaks. I can replace it with an uglier sub-`match`, of course. Still, I am checking if adding this nightly feature instead would be ok?

File: riscv/src/compiler.rs [hunk 8]
```diff
@@ -202,6 +206,25 @@ fn replace_dynamic_label_reference(
...
+                .collect();
+
+            if stub_names.contains(&label.as_str()) {
+                to_delete.insert(i);        // for the label
+                to_delete.insert(i + 1);    // for the `ret` instruction
+            }
+        }
+    }
+    let mut i = 0;
+    statements.retain(|_| (!to_delete.contains(&i), i += 1).0);
```
Comment (Schaeff) [inline riscv/src/compiler.rs:225]: This is a bit obscure to me. How about something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f5a1a35a9d84adb7e81a59b0171636f7) for this whole function?

File: riscv/src/compiler.rs [hunk 9]
```diff
@@ -841,6 +875,17 @@ fn only_if_no_write_to_zero_vec(statements: Vec<String>, reg: Register) -> Vec<S
     }
 }
 
+// let coprocessor_substitutions = vec![("poseidon_coprocessor", "")];
+static COPROCESSOR_SUBSTITUTIONS: &'static [(&'static str, &'static str)] =
+    &[("poseidon_coprocessor", "x10 <=X= poseidon(x10, x11);")];
+
+fn has_coprocessor_substitution(label: &str) -> Option<&'static str> {
```
Comment (Schaeff) [inline riscv/src/compiler.rs:882]: A better name would be `try_to_coprocessor_substitution`. `has_*` should return a boolean.

File: riscv/src/compiler.rs (hunk 7 above)
Comment (Schaeff) [reply riscv/src/compiler.rs:1186]: That's unfortunate but I think we should not add a nightly feature.

File: riscv/src/compiler.rs (hunk 8 above)
Comment (gballet) [reply riscv/src/compiler.rs:225]: I adapted it from something I found in the same file, but sure that would work too;

File: riscv/src/compiler.rs (hunk 7 above)
Comment (gballet) [reply riscv/src/compiler.rs:1185]: you break my heart

File: riscv/src/compiler.rs [hunk 10]
```diff
@@ -705,6 +735,10 @@ fn runtime() -> &'static str {
 
 .globl __rust_alloc_error_handler
 .set __rust_alloc_error_handler, __rg_oom
+
+.globl poseidon_coprocessor
+poseidon_coprocessor:
+    ret
```
Comment (leonardoalt) [inline riscv/src/compiler.rs:741]: Why is this needed?

Comment (gballet) [reply riscv/src/compiler.rs:741]: because otherwise the linker will complain about a missing symbol:

```
The assembly code references an external routine / label that is not available:
runtime_dash_eee54d30a73402ee__poseidon_coprocessor
```

File: riscv/src/compiler.rs (hunk 9 above)
Comment (Schaeff) [reply riscv/src/compiler.rs]: I think so yes!

**Review by Schaeff — APPROVED** [review]

**Comment by Schaeff** [comment]
LGTM! Could you squash?

**Comment by leonardoalt** [comment]
> LGTM! Could you squash?

we do "squash and merge" now
